### DISCUSSION TITLE
Statically define queries accessible to class

### DISCUSSION THREAD

**Discussion by yringler** [LuckyPennySoftware/MediatR#1089] [Ideas]
As soon as IMediator is injected to a class, their is no limit to what the class can do; it's conceivable that the wrong query / notification could be used, and mistakes could be spread around anywhere in a class. There's no definition of capability at the class level.

Alternatively, if you inject several services into a class constructor, there's a single point of reference where
1. You can see something out of place
2. You can see if the class is doing too much

What if there was a way to get those benefits in mediator?

Something like

```csharp
// C# 11, older versions would have to use typeof
[UsesQuery<EngineQuery>, UsesQuery<TireQuery>, UsesQuery<BeltQuery>]
public class CarBuilder(IMediator mediator)
{
}
```

And a custom rosalyn analyzer which could provide design and build time validation?

**Comment by zachpainter77** [comment]
I don't really understand the need for this, myself.. Why wouldn't you just namespace the queries and then only reference the ones you need?
